lib/resty/auto-ssl/jobs/renewal.lua

```diff
@@ -125,12 +127,22 @@ local function renew_check_cert(auto_ssl_instance, storage, domain)
   end
 
   -- If expiry date is known, attempt renewal if it's within 30 days.
+  -- Between 30 and 15 days out, only attempt renewal of a subset of domains (with
+  -- increasing likelihood of renewal being attempted).
   if cert["expiry"] then
     local now = ngx.now()
     if now + (30 * 24 * 60 * 60) < cert["expiry"] then
       ngx.log(ngx.NOTICE, "auto-ssl: expiry date is more than 30 days out, skipping renewal: ", domain)
       renew_check_cert_unlock(domain, storage, local_lock, distributed_lock_value)
       return
+    elseif now + (15 * 24 * 60 * 60) < cert["expiry"] then
+      local rand_value = math.random(cert["expiry"] - (30 * 24 * 60 * 60), cert["expiry"] - (15 * 24 * 60 * 60))
+      local rand_renewal_threshold = now
+      if rand_value < rand_renewal_threshold then
+        ngx.log(ngx.NOTICE, "auto-ssl: expiry date is more than 15 days out, randomly not picked for renewal: ", domain)
+        renew_check_cert_unlock(domain, storage, local_lock, distributed_lock_value)
+        return
+      end
```

Comment:
Is this increasing likelihood approach needed with the other throttle mechanisms provided by `renewals_per_hour` (60/hour)? I guess I'm just trying to think through the different scenarios with this and wondering how these two approaches interact.

If you have a big batch of domains that need to be renewed, would the `renewals_per_hour` logic not eventually get through renewed, just at a slower pace? Or is this logic here an attempt to prioritize the renewals, so that ones expiring sooner have their renewal attempted sooner? Would this mainly be relevant if you expect your renewals to take more than 30 days at the throttled 60/hour rate?

If this is mainly intended to prioritize renewals (but you still want to saturate the 60/hour rate), then I'm wondering if maybe a slightly simpler approach would be to sort the renewals by expiration (instead of the current approach of randomizing the renewal order). That being said, there may be some performance implications of fetching all of those expiration times before renewals occur, so this may not actually be feasible, I'd need to investigate further.

Reply:
Great question about how this interacts with the throttle mechanism!

I implemented this after realizing that we had a bulk of many-thousand of certificates attempting to renew _at the same day_, as this was the date when we migrated them onto `lua-resty-auto-ssl` initially.

While this went on, at some point we exceeded the rate limit, and this caused an issue, as we weren't able to generate certificates for newly-registered customer domains. This issue likely wouldn't happened if we had set our `renewals_per_hour` more conservatively (we must have used more than 40/h for ad-hoc certificate request).

So the aim of this patch is really just to smoothen out the renewals over many days, anticipating that this will make it less likely that we exceed the rate limit. (Example: I am starting out using `lu-resty-auto-ssl` today and migrate 10k domains onto the system. Before this patch, _all_ of these domains would _simultaneously_ attempted for renewal, a job which would take 7 days to finish, at 60/h. This _could_ be just fine. This patch changes this behavior so that instead the renewals get spread over the next 14 invocations (each of which should only take 12 hours to finish).

In case of the former, we'd be "maxing out" the 60/h for a week straight - a process which would reoccur every two months. In case of the latter, we'd be at 60/h for only half of the time, the rest of which we'd have a little bit more room for ad-hoc certificate requests to exceed 40/h.

Does this make more sense to you after reading this?